### PR TITLE
Support naming instances

### PR DESCRIPTION
**Pull request by rbradford**
This PR implements some basic support for names for instances. These names must be tenant unique and will be used for the hostname inside the instance (and thus the DHCP created DNS entry.)

### CODE REVIEW

File: ciao-controller/instance.go [hunk 1]
```diff
@@ -62,10 +62,22 @@ func isCNCIWorkload(workload *types.Workload) bool {
...
-	volumes []storage.BlockDevice) (*instance, error) {
+	volumes []storage.BlockDevice, name string) (*instance, error) {
 	id := uuid.Generate()
 
-	config, err := newConfig(ctl, workload, id.String(), tenantID, volumes)
+	if name != "" {
+		existingID, err := ctl.ds.ResolveInstance(tenantID, name)
+		if err != nil {
+			return nil, errors.Wrap(err, "error trying to resolve name")
+
```
Comment (rbradford) [inline ciao-controller/instance.go]: Fixed.

File: ciao-controller/instance.go [hunk 2]
```diff
@@ -62,10 +62,22 @@ func isCNCIWorkload(workload *types.Workload) bool {
...
+		}
+
+		if existingID != "" {
```
Comment (rbradford) [inline ciao-controller/instance.go]: I've added a database constraint to enforce this beyond this test.

File: ciao-controller/internal/datastore/datastore.go [hunk 3]
```diff
@@ -2393,3 +2393,25 @@ func (ds *Datastore) GetQuotas(tenantID string) ([]types.QuotaDetails, error) {
...
+	t, ok := ds.tenants[tenantID]
+	if ok {
+		for _, i := range t.instances {
+			if i.Name == name || i.ID == name {
+				return i.ID, nil
+			}
+		}
+	}
+
+	if !ok {
```
Comment (rbradford) [inline ciao-controller/internal/datastore/datastore.go]: Fixed.

File: ciao-controller/command.go [hunk 4]
```diff
@@ -247,6 +255,7 @@ func (c *controller) launchCNCI(tenantID string) error {
 		WorkloadID: workloadID,
 		TenantID:   tenantID,
 		Instances:  1,
+		Name:       "cnci",
```
Comment (kaccardi) [inline ciao-controller/command.go]: will it be a problem to have multiple hosts named "cnci"?

Comment (kaccardi) [inline ciao-controller/command.go]: what if a user decides to name their instance "cnci"? how will we use this hostname?

File: ciao-controller/openstack_compute.go [hunk 5]
```diff
@@ -359,27 +361,29 @@ func (c *controller) CreateServer(tenant string, server compute.CreateServerRequ
 		nInstances = server.Server.MinInstances
 	}
 
+	if server.Server.Name != "" {
+		r := regexp.MustCompile("^[a-z0-9-]*$")
```
Comment (kaccardi) [inline ciao-controller/openstack_compute.go]: do you think we should put a limit on the length for the name as well? wondering if someone could be evil here.

File: ciao-controller/command.go (hunk 4 above)
Comment (rbradford) [inline ciao-controller/command.go]: As names are optional I can avoid this problem by just not setting a name. The CNCI already gets a special hostname in the code anyway.

File: ciao-controller/openstack_compute.go (hunk 5 above)
Comment (rbradford) [inline ciao-controller/openstack_compute.go]: Yes, we should limit to HOST_NAME_MAX - 64.

**Comment by rbradford** [comment]
@kaccardi Addressed your feedback. Thanks!

**Comment by kaccardi** [comment]
LGTM

**Pull request merged**